tensorflow/core/kernels/data/captured_function.cc

```diff
@@ -629,7 +629,9 @@ Status InstantiatedCapturedFunction::RunInstantiated(
   FunctionLibraryRuntime::Options f_opts;
   ScopedStepContainer step_container(
       f_opts.step_id, [this](const string& name) {
-        lib_->device()->resource_manager()->Cleanup(name).IgnoreError();
+        if (!lib_->device()->resource_manager()->IsClearingResource()) {
```

Comment:
What is happening here? I don't understand why cleanup only needs to be called if other resources are not being cleared.

Reply:
@alextp Yes, multiple resources can be cleaned in parallel by calling  [ResourceMgr::Cleanup()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/resource_mgr.cc#L220), which uses `tf_shared_lock`. 

The problem here is that in the above generator_data [case](https://github.com/tensorflow/tensorflow/pull/30102#issue-291325728), `ResourceMgr::Clear()` internally calls `ResourceMgr::Cleanup()` by `GeneratorDatasetOp::Dataset::Iterator::~Iterator()`. As [ResourceMgr::Clear()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/resource_mgr.cc#L110) uses `metex_lock`, it causes the deadlock with `ResourceMgr::Cleanup()`.

The function names in this PR may be confusing. What this PR wants to do is that if `ResourceMgr::Clear()` is running and it internally calls `ResourceMgr::Cleanup()` somehow (e.g. a deconstructor runs some captured instantiated functions, which call `ResourceMgr::Cleanup()` inside), then the call of `ResourceMgr::Cleanup()` will be skipped. In my understanding, when `ResourceMgr::Clear()` is running, the resources can not be added/allocated, so it may be ok to skip the call of `ResourceMgr::Cleanup()`.

Reply:
Ok I think the thing I do not understand is why do we need to call cleanup here at all, instead of just removing a particular resource.

Reply:
@alextp Could you please give more explanation about "just removing a particular resource"? 

`CleanUp()` is called here to remove the resources used when running the instantiated function. It is necessary to call `CleanUp()` for the normal cases (where the iterator of generator dataset reaches the end of sequence), since `ResourceMgr::Clear()` will not be called. For the special case mentioned above (the iterator does not reach the end of sequence and `ResourceMgr::Clear()` is called by session close()), we do not need to call `ResourceMgr::Cleanup()`. 

Basically, this PR wants to make `InstantiatedCapturedFunction::RunInstantiated()` handle this special case and also work on normal cases. I'm not sure if this kind of deadlock will happen somewhere else. If the change of ResourceMgr here does not make too much sense to other codes, I can try another approach. For example, `InstantiatedCapturedFunction::RunInstantiated(
    const std::vector<Tensor>& args, std::vector<Tensor>* rets)` is only used by `GeneratorDatasetOp::Dataset`, so we can change this function to handle the special case with no need of changing `ResourceMgr`.

Reply:
So cleanup takes a container name. Why do we have to clear an entire container instead of deleting a few specific resources?

This is what confuses me here.

Reply:
In my understanding, the container here is used for per-step resources. Once this step finishes, all resources in this step need to be removed.

Reply:
Do we still need the per-step container?

Reply:
If we do, the right fix is to store somewhere that we need to run a cleanup operation and run it after we release the mutex.

Reply:
@alextp Thanks for the explanation! Got it now and I will try to implement the fix you suggest.
@jsimsa may know whether we need the per-step container here.

Reply:
Looking at the code base, `ScopedStepContainer` is used by ops that create resources whose life-time is scoped within a step and the method that is passed to the `ScopedStepContainer` is used to cleanup these resources. However, the argument to the cleanup method is that name of the container (so that all resources it contains are cleaned up), not a resource handle.

It seems that the `ResourceMgr::Clear` does not expect that a destructor it triggers could in turn call methods on the `ResourceMgr` object.

I think a better way to fix this problem would be to make modify `Clear` to move the contents of `containers_` to a temporary local variable under the mutex (and then perform that actual unref-ing and deallocation without holding the lock).

Reply:
+1. Thanks @jsimsa! Will work it.